pkg/build/registry/build/strategy.go

```diff
@@ -0,0 +1,125 @@
+package build
+
+import (
+	"fmt"
+	"time"
+
+	"code.google.com/p/go-uuid/uuid"
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/generic"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/fielderrors"
+
+	"github.com/openshift/origin/pkg/build/api"
+	"github.com/openshift/origin/pkg/build/api/validation"
+	buildutil "github.com/openshift/origin/pkg/build/util"
+)
+
+// strategy implements behavior for nodes
+type strategy struct {
+	runtime.ObjectTyper
+}
+
+// Decorator is used to compute duration of a build since its not stored in etcd yet
+var Decorator = func(obj runtime.Object) error {
+	build, ok := obj.(*api.Build)
+	if !ok {
+		return errors.NewBadRequest(fmt.Sprintf("not a build: %v", build))
+	}
+	if build.StartTimestamp == nil {
+		build.Duration = time.Duration(0)
+	} else {
+		completionTimestamp := build.CompletionTimestamp
+		if completionTimestamp == nil {
+			dummy := util.Now()
+			completionTimestamp = &dummy
+		}
+		build.Duration = completionTimestamp.Rfc3339Copy().Time.Sub(build.StartTimestamp.Rfc3339Copy().Time)
+	}
+	return nil
+}
+
+// Strategy is the default logic that applies when creating and updating Build objects.
+var Strategy = strategy{kapi.Scheme}
+
+func (strategy) NamespaceScoped() bool {
+	return true
+}
+
+// AllowCreateOnUpdate is false for Build objects.
+func (strategy) AllowCreateOnUpdate() bool {
+	return false
+}
+
+func (strategy) GenerateName(base string) string {
+	return base
+}
+
+// PrepareForCreate clears fields that are not allowed to be set by end users on creation.
+func (strategy) PrepareForCreate(obj runtime.Object) {
+	build := obj.(*api.Build)
+
+	// TODO this is preserving the old behavior, ideally we could do this in generateName but generateName
+	// does not have access to the full build object
```

Comment:
We don't ever use name generation this way.  We always give a name directly.  It should be safe to drop the legacy behavior (@soltysh correct me if I'm wrong)

----- Original Message -----
> > +
> > +// AllowCreateOnUpdate is false for Build objects.
> > +func (strategy) AllowCreateOnUpdate() bool {
> > +	return false
> > +}
> > +
> > +func (strategy) GenerateName(base string) string {
> > +	return base
> > +}
> > +
> > +// PrepareForCreate clears fields that are not allowed to be set by end
> > users on creation.
> > +func (strategy) PrepareForCreate(obj runtime.Object) {
> > +	build := obj.(*api.Build)
> > +
> > +	// TODO this is preserving the old behavior, ideally we could do this in
> > generateName but generateName
> > +	// does not have access to the full build object
> 
> This was a comment that I put in place per our discussion yesterday and
> forgot to go update it.
> 
> I was not sure if there was an expected behavior that the build config label
> needed to be used as input to any name generation.  I would prefer to keep
> this here for now and do any name generation changes in a follow-up if that
> is fine.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2204/files#r30251347